### ISSUE TITLE
Improve Detailed Tablet Operations in Monitor

### ISSUE THREAD

**Issue by milleruntime** [apache/accumulo#1940]
The monitor has a page for details about a tserver that also displays information about tablets on that tserver. This information could be useful but for security purposes the actual tablets (split points) have been obscured (see screenshot). The information isn't very useful if you can't pin point the tablet that may be having problems. AFAIK there is no way for an admin to decrypt the tablets. Since splits don't have to be sensitive, I think it would be reasonable to have a property to disable the tablet obscuring in the monitor. 

Another option would be to write a command line utility to allow admins with the proper permissions to decrypt the tablets but that is a lot more involved and would be very cumbersome.

**Screenshots**
![Screenshot_2021-02-19_12-09-06](https://user-images.githubusercontent.com/11872539/108538059-5de28600-72ac-11eb-933c-a11abfd30d0c.png)

**Comment by ctubbsii** [comment]
I thought we had a command in the shell to list tablets in a de-obfuscated way, for an authenticated user, so they can cross-reference with the monitor.

**Comment by ctubbsii** [comment]
Without authentication, I don't think we should de-obfuscate the monitor... and I don't think it's worth baking in authentication to the monitoring service. It'd probably be better to have an authenticated TUI to list the detailed information that the monitor is showing instead.

**Comment by milleruntime** [comment]
> I thought we had a command in the shell to list tablets in a de-obfuscated way, for an authenticated user, so they can cross-reference with the monitor.

I added a `listtablets` command but that is separate from the monitor and you will have no way to associate it with the information on the monitor.

**Comment by milleruntime** [comment]
> and I don't think it's worth baking in authentication to the monitoring service. 

I agree. This issue wasn't for that.

> It'd probably be better to have an authenticated TUI to list the detailed information that the monitor is showing instead.

TUI? Trusted User Interface?

**Comment by ctubbsii** [comment]
TUI = https://en.wikipedia.org/wiki/Text-based_user_interface

**Comment by ctubbsii** [comment]
> > I thought we had a command in the shell to list tablets in a de-obfuscated way, for an authenticated user, so they can cross-reference with the monitor.
> 
> I added a `listtablets` command but that is separate from the monitor and you will have no way to associate it with the information on the monitor.

No, the `getsplits` command has a verbose option to show the obfuscated form that would appear on the monitor.

I'm not sure what `listtablets` does exactly, but if it overlaps with `getsplits`, it might be worth consolidating its features into a `getsplits` option, since that command existed already, and it might be somewhat redundant.

**Comment by EdColeman** [comment]
This comment may be late in the game - but:

`Since splits don't have to be sensitive, I think it would be reasonable to have a property to disable the tablet obscuring in the monitor.`

Split points can be very sensitive and should only available to admin users.

**Comment by milleruntime** [comment]
> Split points can be very sensitive and should only available to admin users.

I understand they _can_ be very sensitive. My point is that they don't have to be. They can also be mundane. 

Do you think you or anyone would ever use a separate command line utility (like getsplits) to translate the encrypted tablet text on the monitor? It seems like a very cumbersome task and if you are already on the command line, there are probably better ways of finding out if there are issues with a tablet.

**Comment by milleruntime** [comment]
> No, the getsplits command has a verbose option to show the obfuscated form that would appear on the monitor.

This does work. At the very least I think we should put something on the monitor so a user knows that `getsplits -v` will provide the same information.

**Comment by ctubbsii** [comment]
> Do you think you or anyone would ever use a separate command line utility (like getsplits) to translate the encrypted tablet text on the monitor?

Yes, they do.

> It seems like a very cumbersome task and if you are already on the command line, there are probably better ways of finding out if there are issues with a tablet.

It may be cumbersome, but it was done intentionally in order to avoid leaking user data. As far as I can tell, nobody has developed a better way to securely identify tablet issues on the monitor, and for an administrator to associate them with the information they have access to in the server logs. This is the best, and only, way, we have come up with.

**Comment by milleruntime** [comment]
What about a simple number sequence? Something generated like `Tablet-1`, `Tablet-2`? It would be the same amount of work to run the getsplits command to translate them but they would be at least human readable.

**Comment by ctubbsii** [comment]
You can't associate "Tablet-1" with a specific split point, unless you track the entire unique history of split points. That's because splits change over time. It doesn't need to be reader-friendly... it just needs to be uniquely associable to a specific tablet for matching up to server logs.

**Comment by milleruntime** [comment]
> You can't associate "Tablet-1" with a specific split point, unless you track the entire unique history of split points. That's because splits change over time. 

It doesn't matter if they change because you can just refresh the page and run the command again. It doesn't need to be a permanent unique association, just a temporary one from the time the user looks at the page and runs the command.

> It` doesn't need to be reader-friendly... it just needs to be uniquely associable to a specific tablet for matching up to server logs.

I disagree. The hashes we generate don't get printed in the logs only on the webpage or run from the `getsplits` command. There is currently no way to search for the tablet hash.

**Comment by EdColeman** [comment]
I think I'm agreeing more with Christopher's characterization? 

tl;dr

This is a long way around to expressing that the split info in the monitor seems of marginal utility to me and should never be display directly by our monitor - if someone needs that, then they need to build their own and handle user authentication and authorization to meet their own requirements.

I'm only relating my recollection of usages - what I have trouble extracting is what could be useful to non-
Accumulo admins. For the most part, if you are dealing with splits, then you are fairly deep into user application implementation and usage details, not something that a typical system admin or operator would find useful. Although in practice, maybe that's not the target audience for these details in the monitor anyway - I'm unsure.

I do not recall ever using the splits as displayed on the monitor.  If there was an issue or question I would normally be examining the current system state with shell commands.  For general state, I would default to scanning the metadata table because it was generally not the split that I'd be looking for but rather where it was hosted or other info (number of files, file sizes, number of entities...).  I only recall using getsplits when doing things like creating batch operations - say to create compactions or merges that would would work on sections of the table at a time - then it's usually piping getsplits into awk to generate a shell command file to run the actual commands. The only times that I recall the actual split value having any usefulness was translating back to application developers where they should consider spending time to generate better row ids.

If the Detailed Tablet Operations tab displayed tablet id and the current assigned server I think I'd have the same info that I needed to start looking deeper - it would still be examining the metadata, but that would be a direct way to locate things, and that would be easier than decoding the hash.  Displaying the server would help with another class of problems when multiple tablets, possibly from different tables are hosted on one server and interact in less than optimal ways.  

To answer the question, if displaying the splits would ever be useful?
* Maybe during development to verify an application is actually generating things as you expect directly. If I was using something like uno and just developing a prototype app - Maybe it would save a few steps? But again I would just as likely be using the shell and already have one open...
* Only if the row id were something generic - say shards based on dates, or some other generic bins - that could allow examining hotspots directly - "say, why is 1970-01-01 00:00 being slammed with ingest,..."  But this would only matter if operators could then act on that info - otherwise it should be straight forward for Accumulo admins to pull that info out anyway.

But, for any "user" splits that are based on the data - they should be protected to at least the sensitivity of the assigned column-visibility of the row-id.  In my opinion, that level of authorization is just something that our monitor should not be expected to handle - that specialization would need to be supplied by users to fit their requirements.  

This would also apply to my shard example above - say the shard info is generic - but the actual rows ids are something like [shard-id]-[my-sensitive-data] if things work well, the shard never splits - and no harm in displaying that - but if that shard ever split - the resulting split point would no longer qualify as generic - it's now sensitive to the column visibility of the rows picked for the split.  And, even if you tried to use the column visibility, you still have an issue - say 99.999% of your data is labeled "anyone" but a few rows are "jane-bond-only" - well, my luck is that it would spilt on Jane's row and then only should be displayed if she was the one looking at the monitor. (and maybe Q, but not sure of the structure in the Bond universe ;-) )

**Comment by ctubbsii** [comment]
> It doesn't need to be a permanent unique association, just a temporary one from the time the user looks at the page and runs the command.

No. It needs to be a permanent unique association, because an active table that is constantly splitting can make association nearly impossible, certainly extremely inconvenient. Additionally, admins often need to associate monitor information with historical log information.

> The hashes we generate don't get printed in the logs only on the webpage or run from the `getsplits` command. There is currently no way to search for the tablet hash.

The typical administrator workflow is:

1. See something funny on monitor page,
2. Identify split using shell,
3. Search logs for activity using split.

The list of splits gathered from any of these sources could be disjoint in time. While there might be information gaps if you don't get the information at the right time to cross-reference, there is no risk of confusion, because there's a unique mapping. With a non-unique mapping, you not only have to deal with information gaps while troubleshooting, you also have to deal with red herrings because a tablet identifier might map to multiple different tablets over time.

One way we could improve this situation is to use the encoded version in the logs as well, so you can cross-reference with the logs directly, instead of needing to translate using the shell, especially since the shell is limited to showing only the current mapping, and not historical mappings that the logs could contain.

One thing we don't want to do though, is use a reversible hash or obfuscation method, or use plain text for the monitor, because we really don't want to leak user data, even if it is safe to do so in some cases. It just doesn't seem to be worth the risk. Also, this data has to be encoded because it could be binary, so the monitor can't even print unencoded/plaint-text split points in the general case anyway.

**Comment by milleruntime** [comment]
> If the Detailed Tablet Operations tab displayed tablet id and the current assigned server I think I'd have the same info that I needed to start looking deeper

The Tablet details are on the page of their current tablet server. Clicking on the tserver address is typically how you get to the page. Every tablet would list the current server of the page you are on so it wouldn't make sense to display the current assigned server.

**Comment by milleruntime** [comment]
I saw this happen during a failure:
<pre>
[INFO] Running org.apache.accumulo.tserver.memory.LargestFirstMemoryManagerTest
[INFO] Running org.apache.accumulo.tserver.TabletResourceManagerTest
[ERROR] WARNING: An illegal reflective access operation has occurred
[ERROR] WARNING: Illegal reflective access by org.easymock.cglib.core.ReflectUtils$1 (file:/home/runner/.m2/repository/org/easymock/easymock/4.2/easymock-4.2.jar) to method java.lang.ClassLoader.defineClass(java.lang.String,byte[],int,int,java.security.ProtectionDomain)
[ERROR] WARNING: Please consider reporting this to the maintainers of org.easymock.cglib.core.ReflectUtils$1
[ERROR] WARNING: Use --illegal-access=warn to enable warnings of further illegal reflective access operations
[ERROR] WARNING: All illegal access operations will be denied in a future release
[ERROR] WARNING: An illegal reflective access operation has occurred
[ERROR] WARNING: Illegal reflective access by org.easymock.cglib.core.ReflectUtils$1 (file:/home/runner/.m2/repository/org/easymock/easymock/4.2/easymock-4.2.jar) to method java.lang.ClassLoader.defineClass(java.lang.String,byte[],int,int,java.security.ProtectionDomain)
[ERROR] WARNING: Please consider reporting this to the maintainers of org.easymock.cglib.core.ReflectUtils$1
[ERROR] WARNING: Use --illegal-access=warn to enable warnings of further illegal reflective access operations
[ERROR] WARNING: All illegal access operations will be denied in a future release
[INFO] Tests run: 1, Failures: 0, Errors: 0, Skipped: 0, Time elapsed: 1.674 s - in org.apache.accumulo.tserver.TabletResourceManagerTest
[ERROR] Tests run: 2, Failures: 1, Errors: 0, Skipped: 0, Time elapsed: 2.007 s <<< FAILURE! - in org.apache.accumulo.tserver.memory.LargestFirstMemoryManagerTest
[ERROR] org.apache.accumulo.tserver.memory.LargestFirstMemoryManagerTest.test  Time elapsed: 1.721 s  <<< FAILURE!
java.lang.AssertionError: expected:<1;f<> but was:<1;e<>
	at org.apache.accumulo.tserver.memory.LargestFirstMemoryManagerTest.test(LargestFirstMemoryManagerTest.java:118)
</pre>

I don't know what is going on but it appears that multiple tests are running at the same time and the other test completes before the LargestFirstMemoryManagerTest.
